src/hotspot/cpu/x86/assembler_x86.cpp

```diff
@@ -1917,69 +1984,85 @@ void Assembler::cpuid() {
-    // Note:
-    // http://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-instruction-set-reference-manual-325383.pdf
-    //
-    // Page B - 72   Vol. 2C says
-    // qwreg2 to qwreg            1111 0010 : 0100 1R0B : 0000 1111 : 0011 1000 : 1111 0000 : 11 qwreg1 qwreg2
-    // mem64 to qwreg             1111 0010 : 0100 1R0B : 0000 1111 : 0011 1000 : 1111 0000 : mod qwreg r / m
-    //                                                                            F0!!!
-    // while 3 - 208 Vol. 2A
-    // F2 REX.W 0F 38 F1 / r       CRC32 r64, r / m64             RM         Valid      N.E.Accumulate CRC32 on r / m64.
-    //
-    // the 0 on a last bit is reserved for a different flavor of this instruction :
-    // F2 REX.W 0F 38 F0 / r       CRC32 r64, r / m8              RM         Valid      N.E.Accumulate CRC32 on r / m8.
-    p = REX_W;
-    break;
-  default:
-    assert(0, "Unsupported value for a sizeInBytes argument");
-    break;
+    emit_int8((unsigned char)0xF2);
+    switch (sizeInBytes) {
+    case 1:
+      w = 0;
+      break;
+    case 2:
+    case 4:
+      break;
+    LP64_ONLY(case 8:)
+      // This instruction is not valid in 32 bits
+      // Note:
+      // http://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-instruction-set-reference-manual-325383.pdf
+      //
+      // Page B - 72   Vol. 2C says
+      // qwreg2 to qwreg            1111 0010 : 0100 1R0B : 0000 1111 : 0011 1000 : 1111 0000 : 11 qwreg1 qwreg2
+      // mem64 to qwreg             1111 0010 : 0100 1R0B : 0000 1111 : 0011 1000 : 1111 0000 : mod qwreg r / m
+      //                                                                            F0!!!
+      // while 3 - 208 Vol. 2A
+      // F2 REX.W 0F 38 F1 / r       CRC32 r64, r / m64             RM         Valid      N.E.Accumulate CRC32 on r / m64.
+      //
+      // the 0 on a last bit is reserved for a different flavor of this instruction :
+      // F2 REX.W 0F 38 F0 / r       CRC32 r64, r / m8              RM         Valid      N.E.Accumulate CRC32 on r / m8.
+      p = REX_W;
+      break;
+    default:
+      assert(0, "Unsupported value for a sizeInBytes argument");
+      break;
+    }
+    LP64_ONLY(prefix(crc, v, p);)
+    emit_int32(0x0F,
+               0x38,
+               0xF0 | w,
+               0xC0 | ((crc->encoding() & 0x7) << 3) | (v->encoding() & 7));
   }
-  LP64_ONLY(prefix(crc, v, p);)
-  emit_int32(0x0F,
-             0x38,
-             0xF0 | w,
-             0xC0 | ((crc->encoding() & 0x7) << 3) | (v->encoding() & 7));
 }
 
 void Assembler::crc32(Register crc, Address adr, int8_t sizeInBytes) {
   assert(VM_Version::supports_sse4_2(), "");
   InstructionMark im(this);
-  int8_t w = 0x01;
-  Prefix p = Prefix_EMPTY;
+  if (needs_eevex(crc, adr.base(), adr.index())) {
+    assert(UseAPX, "extended gpr use requires UseAPX and UseAVX > 2");
```

Comment:
Maybe here the "and" makes sense, but not sure.

Reply:
Yes, "and" is technically correct.  APX includes instructions that require AVX 3 (evex) encoding for extended gpr use, together with +UseAPX.